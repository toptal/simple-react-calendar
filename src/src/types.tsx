import * as Luxon from 'luxon'

export type CalendarSelectionI = CalendarSelectionRangeI | CalendarSelectionSingleI

export interface CalendarRangeSelectedDate extends CalendarSelectionRangeI {
  inProgress: boolean
}

export interface CalendarSelectionRangeI {
  start: Luxon.DateTime
  end: Luxon.DateTime
}

export type CalendarSelectionSingleI = Luxon.DateTime

export enum CalendarSelectionTypeEnum {
  start = 'start',
  end = 'end',
}

export enum CalendarTypeEnum {
  single = 'single',
  range = 'range',
}

export enum MonthHeaderButtonTypeEnum {
  next = 'next',
  prev = 'previous',
}

export enum NoticeTypeEnum {
  overlap = 'overlapping_with_disabled',
  disabled = 'disabled_day_click',
}
