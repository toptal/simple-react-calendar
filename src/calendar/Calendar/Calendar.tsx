import React, { FC, Fragment, SyntheticEvent, useCallback, useEffect, useState } from 'react'

import {
  IDate,
  IDateRange,
  IDayOfWeekRenderProps,
  IDayRenderProps,
  IDaysOfWeekRenderProps,
  IMonthHeaderRenderProps,
  IMonthRenderProps,
  INoticeType,
  ISelectionRange,
  IWeekRenderProps,
} from '../../@types'
import { BLOCK_CLASS_NAME, DAYS_OF_WEEK } from './consts'
import {
  getActiveMonth,
  getAsDate,
  getDay,
  getDayEachDay,
  getDayIsBefore,
  getDayIsCurrentMonth,
  getDayIsDisabled,
  getDayIsHighlighted,
  getDayIsNextMonth,
  getDayISOFormatter,
  getDayIsSame,
  getDayIsSelectable,
  getDayIsSelection,
  getDayIsWeekend,
  getHighlighted,
  getIntervalsAsDate,
  getIsRangeOverlap,
  getMonth,
  getMonthFormattedName,
  getMonthWeeks,
  getSelectedAsDate,
  getSelection,
} from './helper'
import Month from './month'
import Notice from './notice'

export type Props = {
  activeMonth: IDate
  blockClassName: string
  daysOfWeek: string[]
  disabledIntervals?: IDateRange[]
  highlightedIntervals?: IDateRange[]
  maxDate: IDate
  minDate: IDate
  minNumberOfWeeks?: number
  mode?: 'range' | 'single'
  onDayEnter?: (day: Date) => void
  onDayDisabledEnter?: (day: Date) => void
  onMonthChange?: (day: Date) => void
  onNoticeChange?: (noticeType?: INoticeType) => void
  onSelect?: (selected: IDate | ISelectionRange) => void
  onSelectionProgress?: (selected: IDate | ISelectionRange) => void
  selectionRangeLimit?: number
  renderDay?: IDayRenderProps
  renderDayOfWeek?: IDayOfWeekRenderProps
  renderDaysOfWeek?: IDaysOfWeekRenderProps
  renderMonth?: IMonthRenderProps
  renderMonthHeader?: IMonthHeaderRenderProps
  renderWeek?: IWeekRenderProps
  selected?: IDate | ISelectionRange
  showDaysOfWeek: boolean
  today: IDate
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6
}

const Calendar: FC<Props> = (props) => {
  const {
    activeMonth: propActiveMonth,
    blockClassName,
    daysOfWeek,
    disabledIntervals: propDisabledIntervals,
    highlightedIntervals: propHighlightedIntervals,
    maxDate: propMaxDate,
    minDate: propMinDate,
    minNumberOfWeeks,
    mode,
    onDayEnter,
    onDayDisabledEnter,
    onMonthChange,
    onNoticeChange,
    onSelect,
    onSelectionProgress,
    selectionRangeLimit,
    renderDay,
    renderDayOfWeek,
    renderDaysOfWeek,
    renderMonth,
    renderMonthHeader,
    renderWeek,
    selected: propSelected,
    showDaysOfWeek,
    today: propToday,
    weekStartsOn,
  } = props
  /**
   * Convert String and Number to Date instance
   */
  const maxDate = getAsDate(propMaxDate)
  const minDate = getAsDate(propMinDate)
  const activeMonth = getAsDate(propActiveMonth)
  const today = getAsDate(propToday)
  const selected = getSelectedAsDate(propSelected)
  const highlightedIntervals = getIntervalsAsDate(propHighlightedIntervals)
  const disabledIntervals = getIntervalsAsDate(propDisabledIntervals)

  /**
   * Hooks
   */
  const [stateSelection, setStateSelection] = useState<Date | ISelectionRange | undefined>(undefined)
  const [stateNoticeType, setNoticeType] = useState<INoticeType | undefined>(undefined)
  const [stateActiveMonth, setStateActiveMonth] = useState<Date>(
    getActiveMonth({selected, activeMonth, mode: mode as 'range' | 'single', today})
  )

  useEffect(() => {
    setStateActiveMonth(getActiveMonth({activeMonth, selected, mode: mode as 'range' | 'single', today}))
    setStateSelection(getSelectedAsDate(propSelected))
  }, [propActiveMonth, today, propSelected, activeMonth, selected, mode])

  const {start, end} = getHighlighted({highlightedIntervals})

  /**
   * Stored Helper Logic Functions
   */
  const useGetSelection = useCallback(
    () => getSelection({selected, onSelectionProgress, stateSelection, mode: mode as 'range' | 'single'}),
    [selected, onSelectionProgress, stateSelection, mode]
  )
  const useGetDayIsHighlighted = useCallback(
    (date: Date) => getDayIsHighlighted({date, startDay: start, endDay: end}),
    [end, start]
  )
  const useGetDayIsSelectable = useCallback(
    (date: Date) => getDayIsSelectable({maxDate, minDate, date, disabledIntervals}),
    [maxDate, disabledIntervals, minDate]
  )
  const useGetDayIsCurrentMonth = useCallback((date: Date) => getDayIsCurrentMonth({date, activeMonth}), [activeMonth])
  const useGetDayIsNextMonth = useCallback((date: Date) => getDayIsNextMonth({date, activeMonth}), [activeMonth])
  const useGetDayIsDisabled = useCallback((date: Date) => getDayIsDisabled({date, disabledIntervals}), [
    disabledIntervals,
  ])
  const useGetDayIsSelected = useCallback(
    (day: Date) =>
      getDayIsSelection({
        day,
        selection: getSelection({selected, onSelectionProgress, stateSelection, mode: mode as 'range' | 'single'}),
      }),
    [mode, onSelectionProgress, selected, stateSelection]
  )
  const useGetDayIsSelectedEnd = useCallback(
    (day: Date) =>
      getDayIsSelection({
        day,
        type: 'end',
        selection: getSelection({selected, onSelectionProgress, stateSelection, mode: mode as 'range' | 'single'}),
      }),
    [mode, onSelectionProgress, selected, stateSelection]
  )
  const useGetDayIsSelectedStart = useCallback(
    (day: Date) =>
      getDayIsSelection({
        day,
        type: 'start',
        selection: getSelection({selected, onSelectionProgress, stateSelection, mode: mode as 'range' | 'single'}),
      }),
    [mode, onSelectionProgress, selected, stateSelection]
  )
  const useGetDayIsToday = useCallback((day: Date) => getDayIsSame({date: today as Date, dateToCompare: day}), [today])
  const useGetDayIsWeekend = useCallback((day: Date) => getDayIsWeekend({day}), [])
  const useGetWeekMonthWeeks = useCallback(
    (activeMonth: Date) => getMonthWeeks({activeMonth, minNumberOfWeeks, weekStartsOn}),
    [minNumberOfWeeks, weekStartsOn]
  )
  const useGetMonthFormattedName = useCallback((day: Date) => getMonthFormattedName(day), [])
  const useGetDayISOFormatter = useCallback((day: Date) => getDayISOFormatter(day), [])
  const useGetDayEachDay = useCallback(({start, end}: IDateRange) => getDayEachDay({start, end}), [])

  /**
   * Event Handlers
   */
  const handleOnSwitchMonth = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLElement>) => {
    const date = getMonth({date: stateActiveMonth, offsetInMonth: Number(calendarValue)})

    onMonthChange && onMonthChange(date)
    setStateActiveMonth(date)
  }

  // const handleOnNoticeChange = (noticeType?: INoticeType) => {
  //   onNoticeChange && onNoticeChange(noticeType)
  //   setNoticeType(noticeType)
  // }

  const handleOnSelectionChange = (event: SyntheticEvent<HTMLButtonElement>) => {
    console.log('>>>> handleOnSelectionChange', event)
    // const {start, end, inProgress} = changes

    // if (onSelect && start && (mode !== 'range' || !inProgress)) {
    //   onSelect(mode === 'single' ? start : {start, end})
    // }

    // if (mode === 'range') {
    //   if (onSelectionProgress) {
    //     onSelectionProgress(changes)
    //   } else {
    //     setStateSelection(inProgress ? {start, end} : undefined)
    //   }
    // }
  }

  const handleOnDayClick = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLButtonElement>) => {
    const day = getAsDate(calendarValue as string)

    onSelect && onSelect(day)

    if (mode === 'range') {
      const {start, inProgress} = stateSelection as ISelectionRange

      if (inProgress) {
        const isDisabledWithin = getRangeIncludesDisabledInterval({start, end: day, disabledIntervals})

        if (isDisabledWithin) {
          setStateSelection({start: undefined, end: undefined, inProgress: false})
          setNoticeType('overlapping_with_disabled')

          return
        } else {
          setStateSelection({...(stateSelection as ISelectionRange), end: day, inProgress: false})
        }
      } else {
        setStateSelection({start: day, inProgress: true, end: undefined})
      }
    } else {
      setStateSelection(day)
    }

    setNoticeType(undefined)
  }

  const handleOnDayMouseEnter = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLButtonElement>) => {
    const day = getAsDate(calendarValue as string)
    const {inProgress, start, end} = stateSelection as ISelectionRange

    onDayEnter && onDayEnter(day)

    if (!inProgress) return

    const isDisabledWithin = getIsRangeOverlap({selection: stateSelection as ISelectionRange, day})

    if (!isDisabledWithin) return

    const dateLimit = getDay({start, offset: selectionRangeLimit})

    if (getDayIsSame({day1: day, day2: end})) {
      if (!selectionRangeLimit || (selectionRangeLimit && getDayIsBefore({day1: day, date2: dateLimit}))) {
      }
    }
  }

  const handleOnDisabledDayClick = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLButtonElement>) => {
    const day = getAsDate(calendarValue as string)

    onDayDisabledHover && onDayDisabledHover(day)
    console.log('>>>> handleOnDisabledDayClick', event)
  }

  const handleOnDisabledDayHover = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLButtonElement>) => {
    const day = getAsDate(calendarValue as string)

    onDayDisabledHover && onDayDisabledHover(day)
    console.log('>>>> handleOnDisabledDayClick', event)
  }

  const children = (
    <Fragment>
      {!!stateNoticeType && <Notice blockClassName={blockClassName} type={stateNoticeType as INoticeType} />}
      <Month
        activeMonth={stateActiveMonth}
        blockClassName={blockClassName}
        customRender={renderMonth}
        daysOfWeek={daysOfWeek}
        showDaysOfWeek={showDaysOfWeek}
        getDayEachDay={useGetDayEachDay}
        getDayIsCurrentMonth={useGetDayIsCurrentMonth}
        getDayIsDisabled={useGetDayIsDisabled}
        getDayIsHighlighted={useGetDayIsHighlighted}
        getDayIsNextMonth={useGetDayIsNextMonth}
        getDayISOFormatter={useGetDayISOFormatter}
        getDayIsSelectable={useGetDayIsSelectable}
        getDayIsSelected={useGetDayIsSelected}
        getDayIsSelectedEnd={useGetDayIsSelectedEnd}
        getDayIsSelectedStart={useGetDayIsSelectedStart}
        getDayIsToday={useGetDayIsToday}
        getDayIsWeekend={useGetDayIsWeekend}
        getWeekMonthWeeks={useGetWeekMonthWeeks}
        handleOnDayClick={handleOnDayClick}
        handleOnDayMouseEnter={handleOnDayMouseEnter}
        handleOnDisabledDayClick={handleOnDisabledDayClick}
        renderDay={renderDay}
        renderDayOfWeek={renderDayOfWeek}
        renderDaysOfWeek={renderDaysOfWeek}
        renderWeek={renderWeek}
        weekStartsOn={weekStartsOn}
      />
    </Fragment>
  )

  return <div className={blockClassName}>{children}</div>
}

Calendar.defaultProps = {
  blockClassName: BLOCK_CLASS_NAME,
  daysOfWeek: DAYS_OF_WEEK,
  mode: 'single',
  showDaysOfWeek: true,
  today: new Date(),
  weekStartsOn: 1,
}

Calendar.displayName = 'Calendar'

export default Calendar
