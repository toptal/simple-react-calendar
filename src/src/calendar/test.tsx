import {shallow} from 'enzyme'
import * as Luxon from 'luxon'
import * as React from 'react'

import Calendar, {CalendarProps} from '.'
import {CalendarSelectionTypeEnum, CalendarTypeEnum, NoticeTypeEnum} from '../types'

const date = Luxon.DateTime.fromJSDate(new Date(2015, 6, 1))

describe('<Calendar />', () => {
  let wrapper, props, instance

  beforeEach(() => {
    props = getProps({activeMonth: date})
    wrapper = shallow(<Calendar {...props} />)
    instance = wrapper.instance()
  })

  describe('#componentWillReceiveProps', () => {
    context('when nextProps `activeMonth` is defined', () => {
      context('when nextProps `activeMonth` and prop `activeMonth` is different', () => {
        it('sets state `activeMonth`', () => {
          wrapper.setProps({activeMonth: Luxon.DateTime.fromJSDate(new Date(2015, 9, 1))})

          expect(wrapper.state('activeMonth').toISODate()).toBe(
            Luxon.DateTime.fromJSDate(new Date(2015, 9, 1)).toISODate()
          )
        })
      })
    })

    context('when nextProps `activeMonth` is `undefined`', () => {
      it("doesn't set state", () => {
        wrapper.setProps({disableDaysOfWeek: false})

        expect(wrapper.state('activeMonth').toISODate()).not.toBe(
          Luxon.DateTime.fromJSDate(new Date(2015, 9, 1)).toISODate()
        )
      })
    })
  })

  describe('#getInitialMonth', () => {
    context('when prop `activeMonth` is valid date', () => {
      it('returns `activeMonth`', () => {
        expect(instance.getInitialMonth()).toEqual(date)
      })
    })

    context('when prop `activeMonth` is invalid date', () => {
      it('returns `activeMonth`', () => {
        wrapper.setState({
          activeMonth: Luxon.DateTime.fromISO('2017-02-31'),
          selected: 'abc',
          today: date,
        })

        expect(instance.getInitialMonth()).toEqual(date)
      })
    })
  })

  describe('#getActiveMonth', () => {
    context('when prop `onMonthChange` is defined', () => {
      it('returns prop `activeMonth`', () => {
        expect(instance.getActiveMonth()).toBe(date)
      })
    })

    context('when prop `onMonthChange` is `undefined`', () => {
      it('returns state `activeMonth`', () => {
        wrapper.setState({activeMonth: Luxon.DateTime.fromISO('2015-01-01')})

        expect(instance.getActiveMonth()).toEqual(Luxon.DateTime.fromISO('2015-01-01'))
      })
    })
  })

  describe('#getHighlightedDate', () => {
    context('when prop `highlighted` is `undefined`', () => {
      it('returns object', () => {
        expect(instance.getHighlightedDate()).toEqual({start: null, end: null})
      })
    })

    context('when prop `highlighted` is defined', () => {
      it('returns object', () => {
        props = {
          highlighted: {
            start: date,
            end: Luxon.DateTime.fromISO('2015-01-01'),
          },
        }
        wrapper.setProps(props)

        expect(instance.getHighlightedDate()).toEqual(props.highlighted)
      })

      context('when prop `highlighted.start` is invalid date', () => {
        it('returns object', () => {
          props = {
            highlighted: {
              start: 'abc',
              end: Luxon.DateTime.fromISO('2015-01-01'),
            },
          }
          wrapper.setProps(props)

          expect(instance.getHighlightedDate()).toEqual({start: null, end: null})
        })
      })

      context('when prop `highlighted.end` is invalid date', () => {
        it('returns object', () => {
          props = {
            highlighted: {
              start: date,
              end: 'abc',
            },
          }
          wrapper.setProps(props)

          expect(instance.getHighlightedDate()).toEqual({start: null, end: null})
        })
      })
    })
  })

  describe('#getSelection', () => {
    context('when #getDateSelectionDateStart and #getDateSelectionDateEnd returns valid date', () => {
      it('returns object', () => {
        const start = date
        const end = date.plus({day: 5})
        instance.getDateSelectionDateStart = () => start
        instance.getDateSelectionDateEnd = () => end

        expect(instance.getSelection()).toEqual({start, end})
      })
    })

    context('when #getDateSelectionDateStart and #getDateSelectionDateEnd returns invalid date', () => {
      it('returns object', () => {
        instance.getDateSelectionDateStart = () => date
        instance.getDateSelectionDateEnd = () => 'abc'

        expect(instance.getSelection()).toEqual({start: null, end: null})
      })
    })
  })

  describe('#getDateSelectionDate', () => {
    context("when prop `mode` is 'single'", () => {
      it('returns prop `selected', () => {
        props = {selected: date, mode: CalendarTypeEnum.single}
        wrapper.setProps(props)

        expect(instance.getDateSelectionDate(CalendarSelectionTypeEnum.end)).toBe(props.selected)
      })
    })

    context("when prop `mode` is 'range'", () => {
      const selected = {start: date, end: Luxon.DateTime.fromJSDate(new Date(2015, 7, 1))}
      const selection = {
        start: Luxon.DateTime.fromJSDate(new Date(2017, 6, 1)),
        end: Luxon.DateTime.fromJSDate(new Date(2017, 7, 1)),
      }

      context('when prop `onSelectionProgress` is `undefined`', () => {
        beforeEach(() => {
          props = {selected, mode: CalendarTypeEnum.range}
          wrapper.setProps(props)
        })

        context('when state `selection` is `defined`', () => {
          it('returns state `selection.start`', () => {
            wrapper.setState({selection})

            expect(instance.getDateSelectionDate(CalendarSelectionTypeEnum.start)).toBe(selection.start)
          })
        })

        context('when state `selection` is `undefined`', () => {
          it('returns prop `selected.start`', () => {
            expect(instance.getDateSelectionDate(CalendarSelectionTypeEnum.start)).toBe(props.selected.start)
          })
        })
      })

      context('when prop `onSelectionProgress` is defined', () => {
        it('returns prop `selected.start`', () => {
          props = {selected, mode: 'range', onSelectionProgress: true}
          wrapper.setProps(props)
          wrapper.setState({selection})

          expect(instance.getDateSelectionDate(CalendarSelectionTypeEnum.start)).toBe(props.selected.start)
        })
      })
    })
  })

  describe('#handleOnMonthChange', () => {
    const monthDate = Luxon.DateTime.fromISO('2017-01-01')

    context('when prop `onMonthChange` is defined', () => {
      it('calls prop #onMonthChange', () => {
        const onMonthChange = jest.fn()
        wrapper.setProps({onMonthChange})

        instance.handleOnMonthChange(monthDate)

        expect(onMonthChange).toHaveBeenCalledTimes(1)
        expect(onMonthChange).toHaveBeenCalledWith(monthDate)
      })
    })

    context('when prop `onMonthChange` is `undefined`', () => {
      it('sets `activeMonth` state', () => {
        instance.handleOnMonthChange(monthDate)

        expect(wrapper.state('activeMonth')).toBe(monthDate)
      })
    })
  })

  describe('#handleOnChange', () => {
    const selection = {start: date, end: Luxon.DateTime.fromJSDate(new Date(2015, 7, 1))}

    context("when prop `mode` is 'single'", () => {
      context('when prop `onSelect` is defined', () => {
        context('when selection `start` is defined', () => {
          it('calls prop #onSelect', () => {
            const onSelect = jest.fn()
            wrapper.setProps({onSelect, mode: CalendarTypeEnum.single})
            instance.handleOnChange(selection)

            expect(onSelect).toHaveBeenCalledTimes(1)
            expect(onSelect).toHaveBeenCalledWith(selection.start)
          })
        })
      })
    })

    context("when prop `mode` is 'range'", () => {
      context('when prop `onSelect` is defined', () => {
        it('calls prop #onSelect', () => {
          const onSelect = jest.fn()
          wrapper.setProps({onSelect, mode: CalendarTypeEnum.range})
          instance.handleOnChange(selection)

          expect(onSelect).toHaveBeenCalledTimes(1)
          expect(onSelect).toHaveBeenCalledWith(selection)
        })
      })

      context('when prop `onSelectionProgress` is defined', () => {
        it('calls prop #onSelectionProgress', () => {
          props = {mode: CalendarTypeEnum.range, onSelectionProgress: jest.fn()}
          wrapper.setProps(props)

          instance.handleOnChange(Object.assign(selection, {inProgress: true}))

          expect(props.onSelectionProgress).toHaveBeenCalledTimes(1)
          expect(props.onSelectionProgress).toHaveBeenCalledWith(Object.assign(selection, {inProgress: true}))
        })
      })

      context('when prop `onSelectionProgress` is `undefined`', () => {
        beforeEach(() => {
          props = {mode: 'range'}
          wrapper.setProps(props)
        })

        context('when selection `inProgress` is `true`', () => {
          it('sets state `selection`', () => {
            instance.handleOnChange(Object.assign(selection, {inProgress: true}))

            expect(wrapper.state('selection')).toEqual({
              start: date,
              end: Luxon.DateTime.fromJSDate(new Date(2015, 7, 1)),
            })
          })
        })

        context('when selection `inProgress` is `false`', () => {
          it('sets state `selection`', () => {
            instance.handleOnChange(Object.assign(selection, {inProgress: false}))

            expect(wrapper.state('selection')).toBe(null)
          })
        })
      })
    })
  })

  describe('#handleOnNoticeChange', () => {
    it('sets state', () => {
      expect(wrapper.state('shownNoticeType')).toBe(undefined)

      instance.handleOnNoticeChange(NoticeTypeEnum.disabled)

      expect(wrapper.state('shownNoticeType')).toBe(NoticeTypeEnum.disabled)
    })
  })

  describe('#render', () => {
    it('renders <Calendar />', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})

const getProps = (overrides: Partial<CalendarProps> = {}): CalendarProps => ({
  ...overrides,
})
