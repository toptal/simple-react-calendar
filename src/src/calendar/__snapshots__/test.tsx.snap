// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar /> #render renders <Calendar /> 1`] = `
<div
  className="calendar"
>
  <MonthHeader
    activeMonth={"2015-06-30T22:00:00.000-02:00"}
    blockClassName="calendar"
    onMonthChange={[Function]}
  />
  <Month
    activeMonth={"2015-06-30T22:00:00.000-02:00"}
    blockClassName="calendar"
    daysOfWeek={
      Array [
        "Mon",
        "Tue",
        "Wed",
        "Thu",
        "Fri",
        "Sat",
        "Sun",
      ]
    }
    highlightedEnd={null}
    highlightedStart={null}
    mode="single"
    onChange={[Function]}
    onNoticeChange={[Function]}
    selectedMax={null}
    selectedMin={null}
    today={"1999-12-31T22:00:00.000-02:00"}
  />
</div>
`;
